rfd/0164-scoped-rbac.md

```diff
@@ -0,0 +1,1122 @@
+* It is not scalable, as it’s not always useful to set and updates tags for each individual resource
+* It’s hard or impossible to partition infrastructure with two-dimensional labels, although users can use `env: prod` to mark all resources in the production environment, there is no way to say that `env: lab` is a subset of `env: prod`.
+
+#### Current Roles mapping
+
+In Teleport there are multiple ways to map roles to users: static roles mapping to local users and bots, 
+dynamic mapping to SSO via connectors and on-demand assignment via access lists and access requests.
+
+Most users start with local static mapping and SSO mapping, later graduating to access requests and access lists.
+
+https://goteleport.com/docs/access-controls/sso/#creating-an-authentication-connector
+
+For SSO users, on login, Teleport checks SSO connector, and maps attributes of a user to a list of roles:
+
+```yaml
+# connector.yaml
+kind: saml
+version: v2
+metadata:
+  name: corporate
+spec:
+  attributes_to_roles:
+    - {name: "groups", value: "okta-admin", roles: ["access"]}
+     # regular expressions with capture are also supported. 
+     # the next line instructs Teleport
+     # to assign users to roles `admin-1` if his SAML "group" 
+     # attribute equals 'ssh_admin_1':
+   - { name: "group", value: "^ssh_admin_(.*)$", roles: ["admin-$1"] }
+```
+
+This creates several security and scalability issues:
+
+* Every time administrators need to change roles assignment, they have to update the resource.
+* Every time the Identity Provider changes user’s attributes, the users would have to relogin to get new roles. 
+
+For local users, administrators have to specify roles and traits in the local resource:
+
+```yaml
+kind: user
+version: v2
+metadata:
+  name: joe
+spec:
+  # roles is a list of roles assigned to this user
+  roles:
+  - admin
+  # traits are key, list of values pairs assigned to a user resource.
+  # Traits can be used in role templates as variables.
+  traits:
+    logins:
+    - joe
+    - root
+```
+
+This also creates some challenges, as administrators are forced to update local resources each time they have to assign a user new permission.
+
+### Modifications
+
+Let’s now introduce the missing pieces of the puzzle and review how new roles will simplify or deprecate legacy concepts.
+
+#### Hierarchical Resource Groups
+
+Resource groups are one such missing piece - in most cloud environments, resources are split into hierarchy, 
+for example, host `luna` is a member of a resource group `lab`, in turn a member of environment `prod`, which is in turn a member of a Teleport cluster.
+
+In Teleport, we will make cluster a default root of this hierarchy. Every computing resource by default will be a direct member of a root `cluster` resource group.
+
+One resource can be assigned to multiple resource groups at a time, or none. 
```

Comment:
This is a departure from the Azure model, isn't it?  IIRC, in Azure, each resource belongs to at most 1 resource group. (which matches what I would consider a hierarchy).

If the purpose of having a hierarchy is so that you can delegate control of a particular subset to certain admins, without giving them control over resources outside their particular segment, wouldn't we want to ensure that each resource can only belong to a single resource group?

I'm not sure if this will solve the "poison pill" problem otherwise. If we use labels to assign resources to resource groups, and a resource can belong to multiple resource groups, then we still have the problem that the more strict resource group will always win, right?

Reply:
It's OK to have a resource to belong to multiple resource groups. Majority of the assignments will mirror AWS/Azure/GCP hierarchy with this design, with minority using labels approach. It will be up to administrators to define how many hierarchies each resource will be part of, not up to folks who provision resources as it is now.